### PR TITLE
chore(node): apply lint

### PR DESCRIPTION
**Pull request by SimonSchick**
I changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: Does not apply
- [ ] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

This applies *most* lint rules, except ones that would introduce too many breaking changes.
The main changes here are:
 - Removal of almost all `export` keywords (http2 is a special case)
 - line lengths (ugh)
 - removals of generic that the linter didn't like, this _might_ cause some issues but ideally shouldn't.
 - All values are not `const` instead of `var`, this might be debatable because some may re-write globals, please advise.

This (for now) includes https://github.com/DefinitelyTyped/DefinitelyTyped/pull/29147

### CODE REVIEW

**Comment by SimonSchick** [comment]
Note: Changes caused some other packages to fail, is there any way to test this locally @basarat ?
Notable changes in other packages:
Removed `console` declaration of `meteor` since that appears to conflict.
Removed `undefined` from `ParsedUrlQuery` as according to the docs this is not even possible.
Add cast to grunt as it's being a special snowflake.

**Comment by SimonSchick** [comment]
Another update:
Removed `ServerRequest` and `ClientResponse` as these classes do not exist anymore.
Removed `runInDebugContext` as it was removed.

File: types/node/index.d.ts [hunk 1]
```diff
@@ -1013,36 +1014,35 @@ interface IterableIterator<T> { }
...
     }
 
-    export interface ParseOptions {
+    interface ParseOptions {
         maxKeys?: number;
         decodeURIComponent?: Function;
     }
 
-    interface ParsedUrlQuery { [key: string]: string | string[] | undefined; }
+    interface ParsedUrlQuery { [key: string]: string | string[]; }
```
Comment (Flarna) [inline types/node/index.d.ts]: The reason why `| undefined` was used here is that this allows you to augment `ParsedUrlQuery` like this in your application to easier access properties:
```interface ParsedUrlQuery {
  my_key?: string,
  my_arr_key?: string[]
}
```
if `| undefined` is not there it's only possible to specify mandatory properties.

File: types/node/index.d.ts [hunk 2]
```diff
@@ -1296,24 +1285,24 @@ declare module "http" {
...
-    export function request(url: string | URL, options: RequestOptions, callback?: (res: IncomingMessage) => void): ClientRequest;
-    export function get(options: RequestOptions | string | URL, callback?: (res: IncomingMessage) => void): ClientRequest;
-    export function get(url: string | URL, options: RequestOptions, callback?: (res: IncomingMessage) => void): ClientRequest;
-    export var globalAgent: Agent;
+    interface RequestOptions extends ClientRequestArgs { }
+    function request(options: RequestOptions | string | URL, callback?: (res: IncomingMessage) => void): ClientRequest;
+    function request(url: string | URL, options: RequestOptions, callback?: (res: IncomingMessage) => void): ClientRequest;
+    function get(options: RequestOptions | string | URL, callback?: (res: IncomingMessage) => void): ClientRequest;
+    function get(url: string | URL, options: RequestOptions, callback?: (res: IncomingMessage) => void): ClientRequest;
+    const globalAgent: Agent;
```
Comment (Flarna) [inline types/node/index.d.ts]: `globalAgent` is not `const`.

File: types/node/index.d.ts [hunk 3]
```diff
@@ -252,11 +253,11 @@ interface NodeModule {
     paths: string[];
 }
 
-declare var module: NodeModule;
+declare const module: NodeModule;
 
 // Same as module.exports
-declare var exports: any;
-declare var SlowBuffer: {
+declare const exports: any;
```
Comment (Flarna) [inline types/node/index.d.ts]: is this really const?

File: types/node/index.d.ts [hunk 4]
```diff
@@ -1912,56 +1905,56 @@ declare module "https" {
...
-    export function request(url: string | URL, options: RequestOptions, callback?: (res: http.IncomingMessage) => void): http.ClientRequest;
-    export function get(options: RequestOptions | string | URL, callback?: (res: http.IncomingMessage) => void): http.ClientRequest;
-    export function get(url: string | URL, options: RequestOptions, callback?: (res: http.IncomingMessage) => void): http.ClientRequest;
-    export var globalAgent: Agent;
+    function createServer(options: ServerOptions, requestListener?: (req: http.IncomingMessage, res: http.ServerResponse) => void): Server;
+    function request(options: RequestOptions | string | URL, callback?: (res: http.IncomingMessage) => void): http.ClientRequest;
+    function request(url: string | URL, options: RequestOptions, callback?: (res: http.IncomingMessage) => void): http.ClientRequest;
+    function get(options: RequestOptions | string | URL, callback?: (res: http.IncomingMessage) => void): http.ClientRequest;
+    function get(url: string | URL, options: RequestOptions, callback?: (res: http.IncomingMessage) => void): http.ClientRequest;
+    const globalAgent: Agent;
```
Comment (Flarna) [inline types/node/index.d.ts]: I think this is not const. It may be a valid usecase to modify the global agent in an application.

**Comment by Flarna** [comment]
Quite a big changeset so maybe I haven't overlooked something...
Please note that `inspector.d.ts` is a generated file generated via the scripts in types/node/scripts/generate-inspector therefore I think these scripts should be adapted instead of `inspector.d.ts`.

**Review by jeffmay — APPROVED** [review]

File: types/node/index.d.ts (hunk 2 above)
Comment (SimonSchick) [reply types/node/index.d.ts]: Seems that way, however changing it on the fly could break references.
I will change it to let.

File: types/node/index.d.ts (hunk 1 above)
Comment (SimonSchick) [reply types/node/index.d.ts]:
```
        let result: querystring.ParsedUrlQuery;

        const a: {
            a?: string;
        } = result;
```

works just fine in this case (pulled from test).

If you are talking about augmenting the interface through declaration merging however:
I think it's a really really bad practise to do that as you can have various definitions of query string parameters for all sorts of purposes, you should cast to your desired type instead.

File: types/node/index.d.ts [hunk 5]
```diff
@@ -1188,7 +1182,7 @@ declare module "http" {
     }
 
     // https://github.com/nodejs/node/blob/master/lib/_http_server.js#L108-L256
-    export class ServerResponse extends OutgoingMessage {
```
Comment (isLishude) [inline types/node/index.d.ts]: Why remove `export` here? third party library authors maybe need those.

**Review by islishude — CHANGES_REQUESTED** [review]
Remove so much `export`,I don't think that's the right thing to do. there may be compatibility problems.

File: types/node/index.d.ts (hunk 5 above)
Comment (SimonSchick) [reply types/node/index.d.ts]: Disregard this one, see my other comment, this work perfectly fine, as mentioned, `export` was redundant here.

**Review by islishude — APPROVED** [review]

File: types/node/index.d.ts (hunk 1 above)
Comment (Flarna) [reply types/node/index.d.ts]: See the `http.IncomingHttpHeaders` which holds an index signature on the one hand and named optional properties on the other hand to narrow types for known/specified headers. This works only if `| undefined` is added in the index signature.

Users could do the same by augmenting `ParsedUrlQuery` but I agree that this is different and not really an intended usecase.

File: types/node/index.d.ts (hunk 3 above)
Comment (Flarna) [reply types/node/index.d.ts]: Yes, right. adding properties to `exports` is the usecase, not assigning to it.

**Comment by SimonSchick** [comment]
@Flarna my bad, I had `scripts` hidden in my vscode, some config leak I suppose 😛

**Comment by SimonSchick** [comment]
Update:
Updated generator to produce lint conform code (omit redundant namespacing, wrap doc lines, wrap long method lines).

**Comment by SimonSchick** [comment]
Hey @basarat I think this is ready.

File: types/node/index.d.ts [hunk 6]
```diff
@@ -1014,36 +1014,35 @@ interface IterableIterator<T> { }
...
 
-    interface ParsedUrlQuery { [key: string]: string | string[] | undefined; }
+    interface ParsedUrlQuery { [key: string]: string | string[]; }
 
-    export function stringify<T>(obj: T, sep?: string, eq?: string, options?: StringifyOptions): string;
-    export function parse(str: string, sep?: string, eq?: string, options?: ParseOptions): ParsedUrlQuery;
-    export function parse<T extends {}>(str: string, sep?: string, eq?: string, options?: ParseOptions): T;
-    export function escape(str: string): string;
-    export function unescape(str: string): string;
+    function stringify(obj?: {}, sep?: string, eq?: string, options?: StringifyOptions): string;
```
Comment (Flarna) [inline types/node/index.d.ts]: I think the first argument is mandatory.

Comment (SimonSchick) [reply types/node/index.d.ts]: Forgive the ghetto windows shell:

```
C:\Users\Simon>node -v
v10.10.0

C:\Users\Simon>node
> querystring.stringify(undefined)
''
> querystring.stringify()
''
>
```

**Comment by amcasey** [comment]
@Flarna I think you're still reviewing - please let me know if/when you're satisfied.

**Review by Flarna — APPROVED** [review]

File: types/grunt/grunt-tests.ts [hunk 7]
```diff
@@ -13,7 +13,8 @@ interface MyOptions {
 }
 
 // exports should work same as module.exports
-exports = (grunt: IGrunt) => {
+// assigning exports is an error in node, hence the cast
+(<any>global).exports = (grunt: IGrunt) => {
```
Comment (ThomasdenH) [inline types/grunt/grunt-tests.ts]: Isn't the `as` syntax recommended now?

File: types/node/node-tests.ts [hunk 8]
```diff
@@ -38,10 +38,10 @@ import { Buffer as ImportedBuffer, SlowBuffer as ImportedSlowBuffer } from "buff
 //////////////////////////////////////////////////////////
 /// Global Tests : https://nodejs.org/api/global.html  ///
 //////////////////////////////////////////////////////////
-namespace global_tests {
+{
     {
-        let x: NodeModule;
-        let y: NodeModule;
+        const x: NodeModule = <any> {};
+        const y: NodeModule = <any> {};
```
Comment (ThomasdenH) [inline types/node/node-tests.ts]: Same here. [`as` is recommended instead.](https://basarat.gitbooks.io/typescript/docs/types/type-assertion.html)

File: types/node/node-tests.ts [hunk 9]
```diff
@@ -3928,8 +3924,8 @@ namespace http2_tests {
 
     // Http2Stream
     {
-        let http2Stream: http2.Http2Stream;
-        let duplex: stream.Duplex = http2Stream;
+        const http2Stream: http2.Http2Stream = <any> {};
```
Comment (ThomasdenH) [inline types/node/node-tests.ts]: Casting without `as`.

File: types/node/node-tests.ts [hunk 10]
```diff
@@ -3970,31 +3966,31 @@ namespace http2_tests {
         http2Stream.close(undefined, () => {});
 
         // ClientHttp2Stream
-        let clientHttp2Stream: http2.ClientHttp2Stream;
+        const clientHttp2Stream: http2.ClientHttp2Stream = <any> {};
```
Comment (ThomasdenH) [inline types/node/node-tests.ts]: Casting without `as`.

File: types/node/node-tests.ts [hunk 11]
```diff
@@ -3970,31 +3966,31 @@ namespace http2_tests {
...
         clientHttp2Stream.on('push', (headers: http2.IncomingHttpHeaders, flags: number) => {});
         clientHttp2Stream.on('response', (headers: http2.IncomingHttpHeaders & http2.IncomingHttpStatusHeader, flags: number) => {
             const s: number = headers[':status'];
         });
 
         // ServerHttp2Stream
-        let serverHttp2Stream: http2.ServerHttp2Stream;
-        let headers: http2.OutgoingHttpHeaders;
+        const serverHttp2Stream: http2.ServerHttp2Stream = <any> {};
+        const headers: http2.OutgoingHttpHeaders = {};
```
Comment (ThomasdenH) [inline types/node/node-tests.ts]: Casting without `as`.

File: types/node/node-tests.ts [hunk 12]
```diff
@@ -4146,11 +4142,11 @@ namespace http2_tests {
...
         };
         // tslint:disable-next-line prefer-object-spread (ts2.1 feature)
-        let secureClientSessionOptions: http2.SecureClientSessionOptions = Object.assign({}, clientSessionOptions);
+        const secureClientSessionOptions: http2.SecureClientSessionOptions = Object.assign({}, clientSessionOptions);
         secureClientSessionOptions.ca = '';
-        let onConnectHandler = (session: http2.Http2Session, socket: net.Socket) => {};
+        const onConnectHandler = (session: http2.Http2Session, socket: net.Socket) => {};
 
-        let serverHttp2Session: http2.ServerHttp2Session;
+        const serverHttp2Session: http2.ServerHttp2Session = <any> {};
```
Comment (ThomasdenH) [inline types/node/node-tests.ts]: Casting without `as`.

File: types/oauth/oauth-tests.ts [hunk 13]
```diff
@@ -44,7 +44,7 @@ http.createServer((req, res) => {
     } else if (pLen === 2 && p[1].indexOf('code') === 0) {
         /** Github sends auth code so that access_token can be obtained */
         /** To obtain and parse code='...' from code?code='...' */
-        const qsObj = qs.parse<{[key: string]: string}>(p[1].split('?')[1]);
+        const qsObj = <{ code: string }> qs.parse(p[1].split('?')[1]);
```
Comment (ThomasdenH) [inline types/oauth/oauth-tests.ts]: Casting without `as`.

**Review by ThomasdenH — COMMENTED** [review]
The only comment I have is the usage of `<any>` casting instead of using `as any`. [This is the recommended syntax.](https://basarat.gitbooks.io/typescript/docs/types/type-assertion.html)

**Comment by ThomasdenH** [comment]
Not sure if you saw, but there are still these instances of casting with `<>`;

- types/grunt/grunt-tests.ts:17
- types/oauth/oauth-tests.ts:47

**Comment by SimonSchick** [comment]
@ThomasdenH I appreciate the thoroughness but I don't think we should be this nit-picky about this.
I will gladly follow everything that is standartized in the lint config, maybe you should request a change to be made there.

**Comment by ThomasdenH** [comment]
You're probably right. Still, since the PR is about style anyway I think it's a good idea. Anyway, looks good to me now!

**Review by ThomasdenH — APPROVED** [review]

**Pull request merged**